phasespace/fromdecay/genmultidecay.py

```diff
@@ -0,0 +1,290 @@
+import itertools
+from typing import Callable, Union
+
+import tensorflow as tf
+import tensorflow.experimental.numpy as tnp
+from particle import Particle
+
+from phasespace import GenParticle
+
+from .mass_functions import _DEFAULT_CONVERTER
+
+_MASS_WIDTH_TOLERANCE = 0.01
+_DEFAULT_MASS_FUNC = "relbw"
+
+
+class GenMultiDecay:
+    def __init__(self, gen_particles: list[tuple[float, GenParticle]]):
+        """A container that works like GenParticle that can handle multiple decays. Can be created from.
+
+        Args:
+        gen_particles: All the GenParticles and their corresponding probabilities.
+            The list must be of the format [[probability, GenParticle instance], [probability, ...
+        """
+        self.gen_particles = gen_particles
+
+    @classmethod
+    def from_dict(
+        cls,
+        dec_dict: dict,
+        mass_converter: dict[str, Callable] = None,
+        tolerance: float = _MASS_WIDTH_TOLERANCE,
+    ):
+        """Create a GenMultiDecay instance from a dict in the decaylanguage format.
+
+        Args:
+            dec_dict:
+                The input dict from which the GenMultiDecay object will be created from.
+                A dec_dict has the same structure as the dicts used in DecayLanguage.
+            mass_converter: A dict with mass function names and their corresponding mass functions.
+                These functions should take the particle mass and the mass width as inputs
+                and return a mass function that phasespace can understand.
+                This dict will be combined with the predefined mass functions in this package.
+                TODO more docs here
+            tolerance: Minimum mass width of the particle to use a mass function instead of
+                assuming the mass to be constant.
+
+        Returns:
+            The created GenMultiDecay object.
+
+        Examples:
+            DecayLanguage is usually used to create a dict that can be understood by GenMultiDecay.
+            >>> from decaylanguage import DecFileParser
+            >>> from phasespace.fromdecay import GenMultiDecay
+
+            Parse a .dec file to create a DecayLanguage dict describing a D*+ particle
+            that can decay in 2 different ways: D*+ -> D0(->K- pi+) pi+ or D*+ -> D+ gamma.
+
+            >>> parser = DecFileParser('example_decays.dec')
+            >>> parser.parse()
+            >>> dst_chain = parser.build_decay_chains("D*+")
+            >>> dst_chain
+            {'D*+': [{'bf': 0.984,
+                'fs': [{'D0': [{'bf': 1.0,
+                        'fs': ['K-', 'pi+']}]},
+                    'pi+']},
+                {'bf': 0.016,
+                 'fs': ['D+', 'gamma']}]}
+
+            If the D0 particle should have a mass distribution of a gaussian when it decays into K- and pi+
+            a `zfit` parameter can be added to its decay dict:
+            >>> dst_chain["D*+"][0]["fs"][0]["D0"][0]["zfit"] = "gauss"

```

Comment:
This works but is of course rather cumbersome. Also, it won't scale well at all for more complex decay chains. In practice I would always define the zfit parameter via something like this example I posted at some point in this review:
```
from decaylanguage import DecayMode, DecayChain
dm1 = DecayMode(1, 'K- pi+ pi+ pi0', model='PHSP', zfit="rel-BW")
dm2 = DecayMode(1, 'gamma gamma')
dc = DecayChain('D+', {'D+':dm1, 'pi0':dm2})
```
(It is not the same as your example but the point I want to make is obvious.)

This makes me think that it may be rather relevant to enhance this constructor to provide the ability to specify a map between particle and model. In your case I could simply have an argument of the kind `particle_model_map={"D0":"gauss"}` and the step above would be done internally, which is so much better for the user. Of course this would imply that all D0 get the same model, and one would need to define aliases if wanting to differentiate 2 D0's, but that's no different to what decay files do, so not an issue at all,

You may want to delay this to a follow-up MR. Dunno, up to you.

Reply:
I would probably prefer this to be in a follow-up pull request.

Reply:
As said, fine by me. Best to have this in asap and then work on enhancements. I'm pretty sure @jonas-eschle will agree.

So, add this to a to-do list :-).

Reply:
Yes, I agree, that sounds reasonable!